tests/unit/helpers.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""helpers for the unit test."""
+
+import contextlib
+import typing
+import unittest.mock
+
+import ops
+from ops.model import Container
+from ops.testing import Harness
+
+from charm import DiscourseCharm
+
+DATABASE_NAME = "discourse"
+
+
+@contextlib.contextmanager
+def patch_exec(fail: bool = False) -> typing.Generator[unittest.mock.Mock, None, None]:
+    """Patch the ops.model.Container.exec method.
+
+    When fail argument is true, the execution will fail.
+
+    Yields:
+        Mock for the exec method.
+    """
+    exec_process_mock = unittest.mock.MagicMock()
+    if not fail:
+        exec_process_mock.wait_output = unittest.mock.MagicMock(return_value=("", ""))
+    else:
+        exec_process_mock.wait_output = unittest.mock.Mock()
+        exec_process_mock.wait_output.side_effect = ops.pebble.ExecError([], 1, "", "")
+    exec_function_mock = unittest.mock.MagicMock(return_value=exec_process_mock)
+    with unittest.mock.patch.multiple(ops.model.Container, exec=exec_function_mock):
+        yield exec_function_mock
+
+
+def start_harness(
+    *,
+    with_postgres: bool = True,
+    with_redis: bool = True,
+    with_ingress: bool = False,
+    with_config: typing.Optional[typing.Dict[str, typing.Any]] = None,
+):
+    """Start a harness discourse charm.
+
+    This is also a workaround for the fact that Harness
+    doesn't reinitialize the charm as expected.
+    Ref: https://github.com/canonical/operator/issues/736
+
+    Args:
+        - with_postgres: should a postgres relation be added
+        - with_redis: should a redis relation be added
+        - with_ingress: should a ingress relation be added
+        - with_config: apply some configuration to the charm
+
+    Returns: a ready to use harness instance
+    """
+    harness = Harness(DiscourseCharm)
+    harness.begin_with_initial_hooks()
+
+    # We catch all exec calls to the container by default
+    harness.handle_exec("discourse", [], result=0)
+
+    if with_postgres:
+        _add_postgres_relation(harness)
+
+    if with_redis:
+        _add_redis_relation(harness)
+        with patch_exec():
```

Comment:
I am a little curious why the `patch_exec` is still need here since the `harness.handle_exec` has called to handle everything above.

Reply:
Forgot to remove that. Fixed